core/commands/repo.go

```diff
@@ -95,3 +100,141 @@ order to reclaim hard disk space.
 		},
 	},
 }
+
+type RepoLsRootsOutput struct {
+	Hash   string
+	Size   uint64
+	Type   string
+	Pinned string
+}
+
+var repoLsRootsCmd = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "Display the top nodes stored locally",
+		ShortDescription: `
+'ipfs repo ls-roots' will display the top-level files or directory
+that are stored locally as well as some of their properties.
+`,
+	},
+
+	Run: func(req cmds.Request, res cmds.Response) {
+		n, err := req.InvocContext().GetNode()
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+
+		// Force pinner flush as we ask the blockstore directly
+		n.Pinning.Flush()
+
+		outChan := make(chan interface{})
+		res.SetOutput((<-chan interface{})(outChan))
+
+		keychan, err := n.Blockstore.AllKeysChan(req.Context())
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+
+		go func() {
+			defer close(outChan)
+			roots := set.NewSimpleBlockSet()
+			childs := set.NewSimpleBlockSet()
+
+			// find the roots of the DAG
+		KeyLoop:
+			for k := range keychan {
+
+				// skip internal pinning node
+				for _, pinnedKey := range n.Pinning.InternalPins() {
+					if pinnedKey == k {
+						continue KeyLoop
+					}
+				}
+
+				node, err := n.DAG.Get(req.Context(), k)
+				if err != nil {
+					res.SetError(err, cmds.ErrNormal)
+					return
+				}
+
+				if !childs.HasKey(k) {
+					roots.AddBlock(k)
+				}
+
+				for _, child := range node.Links {
+					childKey := key.Key(child.Hash)
+					roots.RemoveBlock(childKey)
+					childs.AddBlock(childKey)
+				}
+			}
+
+			for _, k := range roots.GetKeys() {
+				node, err := n.DAG.Get(req.Context(), k)
+				if err != nil {
+					res.SetError(err, cmds.ErrNormal)
+					return
+				}
+
+				unixFSNode, err := unixfs.FromBytes(node.Data)
+				if err != nil {
+					// ignore nodes that are not unixFs
```

Comment:
instead of ignoring them, maybe we could just put their type as 'Unknown' or something?